File: scripts/verify_lottie_player.py
from __future__ import annotations

import argparse
from functools import partial
import hashlib
from http.server import SimpleHTTPRequestHandler, ThreadingHTTPServer
import json
from pathlib import Path
import re
import shutil
import struct
import subprocess
import tempfile
import threading
from typing import Any

ROOT = Path(__file__).resolve().parents[1]
DEFAULT_ROOT = ROOT / "runtime" / "lottie"


def sha256_file(path: Path) -> str:
    digest = hashlib.sha256()
    with path.open("rb") as handle:
        for chunk in iter(lambda: handle.read(1024 * 1024), b""):
            digest.update(chunk)
    return digest.hexdigest()


def canonical_json_hash(value: Any) -> str:
    payload = json.dumps(value, sort_keys=True, separators=(",", ":"), ensure_ascii=False)
    return hashlib.sha256(payload.encode("utf-8")).hexdigest()


def png_dimensions(path: Path) -> tuple[int, int]:
    data = path.read_bytes()[:24]
    if len(data) < 24 or data[:8] != b"\x89PNG\r\n\x1a\n":
        raise ValueError(f"not a PNG screenshot: {path}")
    return struct.unpack(">II", data[16:24])


def _attr(dom: str, name: str) -> str | None:
    match = re.search(rf'data-{re.escape(name)}="([^"]*)"', dom)
    return match.group(1) if match else None


def _chrome_command(chrome: str, *, url: str, profile: Path, screenshot: Path | None = None) -> list[str]:
    command = [
        chrome,
        "--headless=new",
        "--no-sandbox",
        "--disable-gpu",
        "--hide-scrollbars",
        "--force-device-scale-factor=1",
        "--window-size=640,360",
        "--virtual-time-budget=1800",
        f"--user-data-dir={profile}",
    ]
    if screenshot is None:
        command.append("--dump-dom")
    else:
        command.append(f"--screenshot={screenshot}")
    command.append(url)
    return command


def _run(command: list[str], *, timeout: int = 30) -> subprocess.CompletedProcess[str]:
    return subprocess.run(command, capture_output=True, text=True, timeout=timeout, check=True)


def verify(
    *,
    root: Path,
    chrome: str,
    player_version: str,
    frames: tuple[int, ...] = (0, 30, 59),
) -> dict[str, Any]:
    animation_path = root / "animation.json"
    contract_path = root / "player_contract.json"
    bundle_path = root / "lottie.min.js"
    integrity_path = root / "npm_integrity.txt"
    for required in (animation_path, contract_path, bundle_path, integrity_path, root / "index.html"):
        if not required.exists():
            raise FileNotFoundError(required)

    document = json.loads(animation_path.read_text(encoding="utf-8"))
    contract = json.loads(contract_path.read_text(encoding="utf-8"))
    canonical_hash = canonical_json_hash(document)
    if canonical_hash != contract.get("document_sha256"):
        raise ValueError("player contract document hash does not match animation.json")
    expected_frames = int(contract.get("expected_frame_count", 0))
    if expected_frames <= 0:
        raise ValueError("player contract frame count missing")
    if any(frame < 0 or frame >= expected_frames for frame in frames):
        raise ValueError("requested verification frame outside animation range")

    class QuietHandler(SimpleHTTPRequestHandler):
        def log_message(self, format: str, *args: object) -> None:  # noqa: A003
            return

    handler = partial(QuietHandler, directory=str(root))
    server = ThreadingHTTPServer(("127.0.0.1", 0), handler)
    thread = threading.Thread(target=server.serve_forever, daemon=True)
    thread.start()
    screenshots = root / "screenshots"
    screenshots.mkdir(exist_ok=True)
    frame_evidence: list[dict[str, Any]] = []
    try:
        port = server.server_address[1]
        for frame in frames:
            url = f"http://127.0.0.1:{port}/index.html?frame={frame}"
            with tempfile.TemporaryDirectory(prefix=f"motion-lottie-dom-{frame}-") as profile:
                dom_result = _run(_chrome_command(chrome, url=url, profile=Path(profile)))
            dom = dom_result.stdout
            if _attr(dom, "ready") != "true":
                raise RuntimeError(f"lottie player did not reach DOMLoaded-ready state at frame {frame}")
            current = int(float(_attr(dom, "current-frame") or "-1"))
            total = int(float(_attr(dom, "total-frames") or "-1"))
            svg_count = int(_attr(dom, "svg-count") or "0")
            requested = int(float(_attr(dom, "requested-frame") or "-1"))
            if requested != frame or current != frame:
                raise RuntimeError(f"lottie frame seek mismatch requested={frame} current={current}")
            if total != expected_frames:
                raise RuntimeError(f"lottie totalFrames mismatch {total}!={expected_frames}")
            if svg_count != 1:
                raise RuntimeError(f"expected exactly one SVG player surface, observed {svg_count}")

            screenshot = screenshots / f"frame_{frame:04d}.png"
            with tempfile.TemporaryDirectory(prefix=f"motion-lottie-shot-{frame}-") as profile:
                _run(_chrome_command(chrome, url=url, profile=Path(profile), screenshot=screenshot))
            width, height = png_dimensions(screenshot)
            if (width, height) != (640, 360):
                raise RuntimeError(f"screenshot dimensions mismatch {(width, height)}")
            frame_evidence.append({
                "frame": frame,
                "current_frame": current,
                "total_frames": total,
                "svg_count": svg_count,
                "png_sha256": sha256_file(screenshot),
                "png_bytes": screenshot.stat().st_size,
            })
    finally:
        server.shutdown()
        server.server_close()
        thread.join(timeout=2)

    unique_frame_hashes = {row["png_sha256"] for row in frame_evidence}
    if len(unique_frame_hashes) < 2:
        raise RuntimeError("physical Lottie frame evidence is visually invariant across sampled frames")

    return {
        "schema": "motion-os.lottie-physical-player/v1",
        "renderer": "lottie-web",
        "player_version": player_version,
        "player_bundle_sha256": sha256_file(bundle_path),
        "npm_dist_integrity": integrity_path.read_text(encoding="utf-8").strip(),
        "document_sha256": canonical_hash,
        "animation_file_sha256": sha256_file(animation_path),
        "expected_frame_count": expected_frames,
        "visual_duration_authority": "frame_count/fps",
        "fps": document["fr"],
        "frame_evidence": frame_evidence,
        "stable_layer_ids": contract.get("stable_layer_ids", []),
        "authority": "RENDERER_EXECUTED",
        "creative_authority": "none",
    }


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument("--root", type=Path, default=DEFAULT_ROOT)
    parser.add_argument("--chrome", default=None)
    parser.add_argument("--player-version", required=True)
    args = parser.parse_args()
    chrome = args.chrome or shutil.which("google-chrome") or shutil.which("chromium") or shutil.which("chromium-browser")
    if not chrome:
        raise SystemExit("Chrome/Chromium executable unavailable")
    evidence = verify(root=args.root.resolve(), chrome=str(chrome), player_version=args.player_version)
    output = args.root / "player_evidence.json"
    output.write_text(json.dumps(evidence, indent=2, sort_keys=True) + "\n", encoding="utf-8")
    print(json.dumps(evidence, indent=2, sort_keys=True))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/verify_lottie_player.mjs
import crypto from 'node:crypto';
import fs from 'node:fs/promises';
import http from 'node:http';
import path from 'node:path';
import process from 'node:process';
import { fileURLToPath } from 'node:url';
import puppeteer from 'puppeteer-core';

const ROOT = path.resolve(path.dirname(fileURLToPath(import.meta.url)), '..');

function parseArgs(argv) {
  const out = {root: path.join(ROOT, 'runtime', 'lottie'), chrome: '', playerVersion: ''};
  for (let i = 0; i < argv.length; i += 1) {
    const key = argv[i];
    if (key === '--root') out.root = path.resolve(argv[++i]);
    else if (key === '--chrome') out.chrome = argv[++i];
    else if (key === '--player-version') out.playerVersion = argv[++i];
    else throw new Error(`unsupported argument: ${key}`);
  }
  if (!out.chrome) throw new Error('--chrome is required');
  if (!out.playerVersion) throw new Error('--player-version is required');
  return out;
}

function sha256(data) {
  return crypto.createHash('sha256').update(data).digest('hex');
}

async function sha256File(filePath) {
  return sha256(await fs.readFile(filePath));
}

function mime(filePath) {
  if (filePath.endsWith('.html')) return 'text/html; charset=utf-8';
  if (filePath.endsWith('.js')) return 'text/javascript; charset=utf-8';
  if (filePath.endsWith('.json')) return 'application/json; charset=utf-8';
  if (filePath.endsWith('.png')) return 'image/png';
  return 'application/octet-stream';
}

async function startServer(root) {
  const rootPath = path.resolve(root);
  const server = http.createServer(async (req, res) => {
    try {
      const url = new URL(req.url, 'http://127.0.0.1');
      if (url.pathname === '/favicon.ico') {
        res.writeHead(204, {'cache-control': 'no-store'});
        res.end();
        return;
      }
      const relative = decodeURIComponent(url.pathname === '/' ? '/index.html' : url.pathname).replace(/^\/+/, '');
      const resolved = path.resolve(rootPath, relative);
      const prefix = `${rootPath}${path.sep}`;
      if (resolved !== rootPath && !resolved.startsWith(prefix)) {
        res.writeHead(403); res.end('forbidden'); return;
      }
      const payload = await fs.readFile(resolved);
      res.writeHead(200, {'content-type': mime(resolved), 'cache-control': 'no-store'});
      res.end(payload);
    } catch (error) {
      res.writeHead(error?.code === 'ENOENT' ? 404 : 500);
      res.end('error');
    }
  });
  await new Promise((resolve) => server.listen(0, '127.0.0.1', resolve));
  return server;
}

async function main() {
  const args = parseArgs(process.argv.slice(2));
  const animationPath = path.join(args.root, 'animation.json');
  const contractPath = path.join(args.root, 'player_contract.json');
  const compilerEvidencePath = path.join(args.root, 'compiler_evidence.json');
  const bundlePath = path.join(args.root, 'lottie.min.js');
  const integrityPath = path.join(args.root, 'npm_integrity.txt');
  const puppeteerIntegrityPath = path.join(args.root, 'puppeteer_integrity.txt');
  const screenshots = path.join(args.root, 'screenshots');
  await fs.mkdir(screenshots, {recursive: true});

  const document = JSON.parse(await fs.readFile(animationPath, 'utf8'));
  const contract = JSON.parse(await fs.readFile(contractPath, 'utf8'));
  const compilerEvidence = JSON.parse(await fs.readFile(compilerEvidencePath, 'utf8'));
  const animationFileHash = await sha256File(animationPath);
  if (animationFileHash !== compilerEvidence.animation_file_sha256) {
    throw new Error('animation.json bytes do not match compiler evidence');
  }
  if (contract.document_sha256 !== compilerEvidence.document_sha256) {
    throw new Error('player contract document hash does not match compiler evidence');
  }
  if (Number(contract.expected_frame_count) !== Number(compilerEvidence.expected_frame_count)) {
    throw new Error('player contract frame count does not match compiler evidence');
  }
  const expectedFrames = Number(contract.expected_frame_count);
  if (!Number.isInteger(expectedFrames) || expectedFrames <= 0) throw new Error('player contract frame count missing');
  const frames = [0, Math.floor(expectedFrames / 2), expectedFrames - 1];

  const server = await startServer(args.root);
  const {port} = server.address();
  const browser = await puppeteer.launch({
    executablePath: args.chrome,
    headless: true,
    args: ['--no-sandbox', '--disable-gpu'],
  });
  const frameEvidence = [];
  try {
    for (const frame of frames) {
      const page = await browser.newPage();
      await page.setViewport({width: 640, height: 360, deviceScaleFactor: 1});
      const consoleErrors = [];
      const pageErrors = [];
      page.on('console', (message) => {
        if (message.type() === 'error') consoleErrors.push(message.text());
      });
      page.on('pageerror', (error) => pageErrors.push(String(error)));
      const url = `http://127.0.0.1:${port}/index.html?frame=${frame}`;
      await page.goto(url, {waitUntil: 'networkidle0', timeout: 15000});
      await page.waitForFunction(
        () => ['true', 'error'].includes(document.documentElement.dataset.ready || ''),
        {timeout: 10000},
      );
      const state = await page.evaluate(() => ({
        ready: document.documentElement.dataset.ready ?? null,
        errorType: document.documentElement.dataset.errorType ?? null,
        requestedFrame: document.documentElement.dataset.requestedFrame ?? null,
        currentFrame: document.documentElement.dataset.currentFrame ?? null,
        totalFrames: document.documentElement.dataset.totalFrames ?? null,
        svgCount: document.documentElement.dataset.svgCount ?? null,
      }));
      if (state.ready !== 'true') {
        throw new Error(`lottie player readiness failure frame=${frame} state=${JSON.stringify(state)} console=${JSON.stringify(consoleErrors)} page=${JSON.stringify(pageErrors)}`);
      }
      const requested = Number(state.requestedFrame);
      const current = Number(state.currentFrame);
      const total = Number(state.totalFrames);
      const svgCount = Number(state.svgCount);
      if (requested !== frame || current !== frame) throw new Error(`lottie frame seek mismatch requested=${frame} current=${current}`);
      if (total !== expectedFrames) throw new Error(`lottie totalFrames mismatch ${total}!=${expectedFrames}`);
      if (svgCount !== 1) throw new Error(`expected exactly one SVG player surface, observed ${svgCount}`);
      if (consoleErrors.length || pageErrors.length) throw new Error(`browser errors at frame ${frame}: ${JSON.stringify({consoleErrors, pageErrors})}`);

      const screenshot = path.join(screenshots, `frame_${String(frame).padStart(4, '0')}.png`);
      await page.screenshot({path: screenshot, type: 'png'});
      const stat = await fs.stat(screenshot);
      frameEvidence.push({
        frame,
        requested_frame: requested,
        current_frame: current,
        total_frames: total,
        svg_count: svgCount,
        png_sha256: await sha256File(screenshot),
        png_bytes: stat.size,
      });
      await page.close();
    }
  } finally {
    await browser.close();
    await new Promise((resolve) => server.close(resolve));
  }

  if (new Set(frameEvidence.map((item) => item.png_sha256)).size < 2) {
    throw new Error('physical Lottie frame evidence is visually invariant across sampled frames');
  }

  const evidence = {
    schema: 'motion-os.lottie-physical-player/v2',
    renderer: 'lottie-web',
    browser_transport: 'puppeteer-core-devtools',
    player_version: args.playerVersion,
    player_bundle_sha256: await sha256File(bundlePath),
    npm_dist_integrity: (await fs.readFile(integrityPath, 'utf8')).trim(),
    puppeteer_dist_integrity: (await fs.readFile(puppeteerIntegrityPath, 'utf8')).trim(),
    compiler_evidence_sha256: await sha256File(compilerEvidencePath),
    document_sha256: contract.document_sha256,
    animation_file_sha256: animationFileHash,
    expected_frame_count: expectedFrames,
    visual_duration_authority: 'frame_count/fps',
    fps: document.fr,
    frame_evidence: frameEvidence,
    stable_layer_ids: contract.stable_layer_ids ?? [],
    authority: 'RENDERER_EXECUTED',
    creative_authority: 'none',
  };
  await fs.writeFile(path.join(args.root, 'player_evidence.json'), `${JSON.stringify(evidence, null, 2)}\n`, 'utf8');
  console.log(JSON.stringify(evidence, null, 2));
}

main().catch((error) => {
  console.error(error?.stack || String(error));
  process.exitCode = 1;
});
